### PR TITLE
Separate attention backends

### PR DESCRIPTION
**Pull request by WoosukKwon**
This PR refactors the attention layer. Specifically, it separates the code paths for Ampere or more recent NVIDIA GPUs (which can directly use FlashAttention) and other GPUs, so that the code for the former becomes much simpler. This PR will also bring some performance improvements for ALiBi models, since we now directly call FlashAttention instead of using xformers in the middle.

### CODE REVIEW

**Comment by WoosukKwon** [comment]
@zhuohan123 What do you think about this design? Please note that while I used `flash_attn` for now, but this will be replaced with FlashInfer.

File: vllm/model_executor/layers/attention/__init__.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+class AttentionFactory:
+
+    @staticmethod
+    def create_attention(*args, **kwargs) -> BaseAttention:
+        if not is_hip() and torch.cuda.get_device_capability()[0] >= 8:
+            from vllm.model_executor.layers.attention.flash import Attention
+            return Attention(*args, **kwargs)
+        else:
+            from vllm.model_executor.layers.attention.non_flash import Attention
```
Comment (zhuohan123) [inline vllm/model_executor/layers/attention/__init__.py:15]: Does `non_flash` mean `xformers + paged attention`? Let's be a bit more clear on the name?

File: vllm/model_executor/layers/attention/__init__.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+            return Attention(*args, **kwargs)
```
Comment (zhuohan123) [inline vllm/model_executor/layers/attention/__init__.py:16]: Let's call the attention class of flash `FlashBasedAttention` and call the non_flash class `NonFlashBasedAttention`? Having the same name can be confusing

File: vllm/model_executor/layers/attention/non_flash.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
+"""Attention layer with Flash and PagedAttention."""
```
Comment (zhuohan123) [inline vllm/model_executor/layers/attention/non_flash.py:1]: Change the comment?

File: vllm/model_executor/layers/attention/paged_attn.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+from vllm._C import ops
+from vllm.model_executor.input_metadata import InputMetadata
+from vllm.model_executor.layers.triton_kernel.prefix_prefill import (
+    context_attention_fwd)
+
+# Should be the same as PARTITION_SIZE in `paged_attention_v2_launcher`.
+_PARTITION_SIZE = 512
+
+
+class PagedAttention:
```
Comment (zhuohan123) [inline vllm/model_executor/layers/attention/paged_attn.py:15]: Call this class `PagedAttentionImpl` or something like `PagedAttentionWrapper`? I think it's better to make it more clear that this is not a PyTorch module class.

File: vllm/model_executor/layers/attention/__init__.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import torch
+
+from vllm.model_executor.layers.attention.base import BaseAttention
+from vllm.utils import is_hip
+
+
+class AttentionFactory:
```
Comment (zhuohan123) [inline vllm/model_executor/layers/attention/__init__.py:7]: I think this `AttentionFactory` is OK but probably not the most ideal. Can we make this class looks more similar to a normal `nn.Module`? I believe this can make the new user who wants to add a new model understand the model code quicker.

**Review by zhuohan123 — COMMENTED** [review]
In general the refactor LGTM. My only small concern is on the learning cost of `AttentionFactory` since it does not completely behave like a torch `nn.module`. I think this can add difficulty for people adding new models.

File: vllm/model_executor/layers/attention/flash.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                    query,
+                    key,
+                    value,
+                    softmax_scale=self.scale,
+                    causal=True,
+                    window_size=self.sliding_window,
+                    alibi_slopes=self.alibi_slopes,
+                )
+            else:
+                # prefix-enabled attention
```
Comment (chenxu2048) [inline vllm/model_executor/layers/attention/flash.py:89]: Prefix-enabled attention and decoding part is the same as that in non_flash.py. Could we move them into BaseAttention? Just like:

```python
class BaseAttention(nn.Module):
    def forward(self, ...):
        if input_metadata.is_prompt:
            if ...:
                self._do_prompt_attention()
            else:
                # prefix-enabled attention
        else:
            # Decoding run.

    def _do_prompt_attention(self):
        # use xformers or flash_attn here
```

File: vllm/model_executor/layers/attention/flash.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            # Decoding run.
+            output = PagedAttention.forward(
+                query,
+                key_cache,
+                value_cache,
+                input_metadata,
+                self.num_kv_heads,
+                self.scale,
+                self.alibi_slopes,
+            )
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/model_executor/layers/attention/flash.py]: Hi @sighingnow, thanks for bringing it up! We are aware of it, but we'd like to just focus on refactoring in this PR.

File: vllm/model_executor/layers/attention/flash.py (hunk 6 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/attention/flash.py]: Hi @chenxu2048, thanks for your input. I intentionally avoided this design since some attention implementation may not follow the structure. For example, an attention kernel may process the prompt attention and prefix-enabled attention together. In terms of flexibility, I think the current structure is preferable.

File: vllm/model_executor/layers/attention/non_flash.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/attention/non_flash.py]: Fixed.

File: vllm/model_executor/layers/attention/__init__.py (hunk 2 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/attention/__init__.py]: Changed into `FlashAttentionBackend` and `XFormersBackend`.

File: vllm/model_executor/layers/attention/__init__.py (hunk 5 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/attention/__init__.py]: I re-organized the code structure. PTAL.

File: vllm/model_executor/layers/attention/flash.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sighingnow) [reply vllm/model_executor/layers/attention/flash.py]: Thanks for clarifying!

**Comment by WoosukKwon** [comment]
@zhuohan123 PTAL. Please note that I intentionally didn't make changes to other models than Llama.

**Review by zhuohan123 — APPROVED** [review]
Thanks! In general LGTM. Will you change all other model files when you merge the PR?

**Comment by pcmoritz** [comment]
Very nice! This is mainly a cosmetic suggestion, but have you considered structuring this as a Python submodule? Like have a folder `attention` in the `vllm/model_executor/layers` folder with the following files:

- `__init__.py`:

```python
from vllm.model_executor.layers.attention.attention import Attention

__all__ = [
    "Attention",
]

```
(this is the public interface)

- `attention.py`: Your `attention.py` file goes here
- `backend/paged_attn.py`: Your `paged_attn.py` file goes here
- `backend/flash_attn.py`: Your `flash_attn.py` file goes here
- `backend/xformers.py`: Your `xformers.py` file goes here

This makes it a little clearer what the interface here is and also follows the structure of the `quantization` and `fused_moe` folder :)

It is interesting that the `paged_attn.py` file is more of a common implementation used by everybody else, I wonder how you think that will evolve in the future :)

**Comment by rkooo567** [comment]
This is an awesome change! When is this planned to be merged to OSS btw? soon?

File: vllm/model_executor/layers/attention_backends/flash_attn.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+        # Reshape the keys and values and store them in the cache.
+        # If key_cache and value_cache are not provided, the new key and value
+        # vectors will not be cached. This happens during the initial memory
+        # profiling run.
+        if key_cache is not None and value_cache is not None:
+            PagedAttentionImpl.reshape_and_cache(key, value, key_cache,
+                                                 value_cache, input_metadata)
+
+        if input_metadata.is_prompt:
```
Comment (Yard1) [inline vllm/model_executor/layers/attention_backends/flash_attn.py:80]: For those classes, I would suggest breaking up `forward` into `forward_prefill` and `forward_decode`, and letting `forward` just pick which one to call.

**Review by rkooo567 — COMMENTED** [review]
Hmm actually I wonder if this is the good level of abstraction. For example, if I'd like to add chunked prefill, I need to reimplement the same thing in both xformer and flash attention backend. 

I wonder if it makes more sense to just support

```python3
class BackendImpl:
    def reshape_and_cache()  # could be renamed to update_kv_cache?
    def prefill()
    def decode()

class PagedAttention:
def forward():
    backendImpl.reshape_and_cache()
    # Each query contains only one of prompt or decoding req
    # It can be changed when chunked prefill is enabled
    assert not (input_metadata.prompt and input_metadata.decoding)
    if input_metadata.prompt > 0:
        backendImpl.prefill()
    if input_metadata.decoding > 0:
        backendImpl.decode()
```

simiilar to flash infer API.
https://docs.flashinfer.ai/

(Flash infer example)

```python3
class FlashInferBackendImpl:
    def __init__(self):
        # setting up decode & prefill wrapper
    def update_kv_cache(self):
        # reshape cache to [num_tokens, num_heads, head_size]
    def prefill():
        # flashinfer.prefill()
    def decode():
        # flashinfer.decode()
```

**Comment by rkooo567** [comment]
Also, separating out profiling run could further simplify the logic

```python3
def forward():
    profiling_run = key_cache is None or value_cache is None
    
    if profiling_run():
        return XformerBackendImpl.prefill()
    
    assert key_cache is not None and value_cache is not None
    ...
    
```

**Comment by WoosukKwon** [comment]
@pcmoritz Thanks for the suggestion! I've changed the directory structure following your advice.

**Comment by WoosukKwon** [comment]
@Yard1 @rkooo567 Thanks for your inputs! Actually, I intentionally avoided the design you proposed to ensure the flexibility in implementing the attention backends. As you pointed out, an attention backend performs 4 tasks: 1) storing the input KV tensors into the KV cache, 2) compute prefills, 3) compute prefills with prefixes, and 4) compute decodes. Currently, the two attention backends (`FlashAttentionBackend` and `XFormersBackend`) have a kernel for each task. However, this may not be necessary true in the future. For example, depending on the kernel implementation, one can compute prefills with and without prefixes (2&3) at the same time. For anther example, an attention kernel in TRT-LLM stores KV cache while computing decodes (1&4). These can be even more complicated if we implement something like [Cascade inference](https://flashinfer.ai/2024/02/02/cascade-inference.html). Hence, I believe we shouldn't fix a certain structure for the attention backends.

**Comment by pythonononer** [comment]
A reasonable explanation！

**Comment by WoosukKwon** [comment]
> Hmm actually I wonder if this is the good level of abstraction. For example, if I'd like to add chunked prefill, I need to reimplement the same thing in both xformer and flash attention backend.

@rkooo567 Actually, this refactoring is indeed for your case. To my knowledge, xformers currently does not support Paged KV cache since it is a pretty recent feature in FlashAttention. If the backends are not separated, we need to add more conditionals inside the unified `Attention` layer.

**Comment by sighingnow** [comment]
> It turns out that adding `flash_attn` to vLLM's dependencies is pretty tricky. Related issue: [Dao-AILab/flash-attention#453](https://github.com/Dao-AILab/flash-attention/issues/453)

The most convenient (and fast) to install flash-attn might be downloading the wheels from flash-attn's release page, rather than `pip3 install flash-attn` as it builds from source, quite time-consuming and error-prone, and, it will prevent vllm being installed on machines without GPU>=A100 (where flash-attn cannot be built).

Checking its existence and its version at runtime in vLLM would be better.

File: vllm/model_executor/layers/attention/flash.py (hunk 6 above)
Comment (chenxu2048) [reply vllm/model_executor/layers/attention/flash.py]: Thanks for your explanation.

**Comment by tdoublep** [comment]
>It turns out that adding flash_attn to vLLM's dependencies is pretty tricky. Related issue: https://github.com/Dao-AILab/flash-attention/issues/453

@WoosukKwon I have also been experimenting a bit with calling flash-attn directly (instead of going xformers), but instead of adding it as an explicit dependency, I found that it was already installed as part of xformers can could be imported like this:
```python
import xformers._C_flashattention as flash
```

In our fork I have been experimenting with calling `flash.varlen_fwd` for prefills, which allows one to remove all padding from the batch. This has some performance advantages in certain scenarios and removes the need to tune `--max-paddings`. Given this new refactoring, could this also be something worth supporting in your view? Is there a reason that vLLM needs to use rectangular padding other than supporting older GPUs where flash isn't supported?

**Comment by rkooo567** [comment]
@tdoublep actually varlen will be more useful after we merge https://github.com/vllm-project/vllm/pull/3236.

File: vllm/model_executor/layers/attention/backends/flash_attn.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+                key_cache,
+                value_cache,
+                input_metadata,
+                self.num_kv_heads,
+                self.scale,
+                self.alibi_slopes,
+            )
+
+        # Reshape the output tensor.
+        return output.view(batch_size, seq_len, hidden_size)
```
Comment (Yard1) [inline vllm/model_executor/layers/attention/backends/flash_attn.py:124]: I would still suggest separating this out into private methods (`_forward_decode`, `_forward_prefill` etc.) so that `forward` can just decide which method to dispatch.

File: vllm/model_executor/layers/attention/ops/paged_attn.py [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+                num_kv_heads,
+                scale,
+                input_metadata.block_tables,
+                input_metadata.context_lens,
+                block_size,
+                input_metadata.max_context_len,
+                alibi_slopes,
+                input_metadata.kv_cache_dtype,
+            )
+        return output
```
Comment (Yard1) [inline vllm/model_executor/layers/attention/ops/paged_attn.py:111]: ditto as in previous comment (`_forward_decode_v1`, `_forward_decode_v2`)

File: vllm/model_executor/layers/attention/backends/flash_attn.py (hunk 9 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/attention/backends/flash_attn.py:124]: > Thanks for your inputs! Actually, I intentionally avoided the design you proposed to ensure the flexibility in implementing the attention backends. As you pointed out, an attention backend performs 4 tasks: 1) storing the input KV tensors into the KV cache, 2) compute prefills, 3) compute prefills with prefixes, and 4) compute decodes. Currently, the two attention backends (FlashAttentionBackend and XFormersBackend) have a kernel for each task. However, this may not be necessary true in the future. For example, depending on the kernel implementation, one can compute prefills with and without prefixes (2&3) at the same time. For anther example, an attention kernel in TRT-LLM stores KV cache while computing decodes (1&4). These can be even more complicated if we implement something like [Cascade inference](https://flashinfer.ai/2024/02/02/cascade-inference.html). Hence, I believe we shouldn't fix a certain structure for the attention backends.

@Yard1 What do you think about this?

Comment (WoosukKwon) [reply vllm/model_executor/layers/attention/backends/flash_attn.py:124]: Got it. First, I believe the current implementation is easy to read; `XFormersBackend` is essentially the same as the current `main` branch and `FlashAttentionBackend` is simpler than that. Particularly for `FlashAttentionBackend`, I believe the implementation in this PR is very easy to understand.

That being said, I do agree that modularizing the backends will make it easy to test them. However, since this PR has already been delayed quite a bit, let's merge the PR and do modularization in the next PR.

File: vllm/model_executor/layers/attention/ops/paged_attn.py (hunk 10 above)
Comment (WoosukKwon) [reply vllm/model_executor/layers/attention/ops/paged_attn.py:111]: ditto. Let's do it in the next PR.

File: vllm/model_executor/layers/attention/backends/flash_attn.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+                    key_cache,
+                    value_cache,
+                    input_metadata,
+                    self.num_heads,
+                    self.num_kv_heads,
+                    self.alibi_slopes,
+                )
+        else:
+            # Decoding run.
+            output = PagedAttentionImpl.forward_decode(
```
Comment (zhaoyang-star) [inline vllm/model_executor/layers/attention/backends/flash_attn.py:113]: I am just curious that why not use `flash_attn_with_kvcache` ? The kernel is faster than paged_attention_kernel. More benchmark details can be found in https://github.com/vllm-project/vllm/pull/2744
